File: 1-StackQueueDeque/javascript/queue.js
// Queue implemented with JS Array, using class

class Queue {
  constructor() {
    this.container = []
  }

  size() {
    return this.container.length
  }

  isEmpty() {
    return !this.container.length
  }

  enqueue(item) {
    // Sequence is reversed: new item is inserted in front
    // and removed from back
    this.container = [item, ...this.container]

    // Could also use
    // this.container.unshift(item)
  }

  dequeue(item) {
    return this.container.pop()
  }

  serialize() {
    return this.container.join("-")
  }
}

const q = new Queue()
console.log(q.size()) // 0
q.enqueue('hello')
q.enqueue('world')
console.log(q.size()) // 2
console.log(q.serialize()) // 'world-hello'
console.log(q.dequeue()) // 'hello'
console.log(q.isEmpty()) // false

File: dataStructures/stackQeueDeque/queue.py
class Queue(object):
    def __init__(self):
        self.container = []

    def __len__(self):
        """
        dunder implementation of size()
        """
        return len(self.container)

    def enqueue(self, item):
        """
        Insert to front
        """
        self.container.insert(0, item)

    def dequeue(self):
        return self.container.pop()

    def is_empty(self):
        return not len(self.container)

if __name__ == "__main__":
    q = Queue()
    q.enqueue(5)
    q.enqueue("Hello")
    q.enqueue("World")
    print(q.dequeue()) # 5
    print(q.dequeue()) # "Hello"
    print(q.dequeue()) # "World"
    print(q.is_empty()) # True

